crates/red_knot_python_semantic/resources/mdtest/terminal_statements.md

```diff
@@ -0,0 +1,620 @@
+# Terminal statements
+
+## Introduction
+
+Terminal statements complicate a naive control-flow analysis.
+
+As a simple example:
+
+```py
+def f(cond: bool) -> str:
+    if cond:
+        x = "test"
+    else:
+        raise ValueError
+    return x
+
+def g(cond: bool):
+    if cond:
+        x = "test"
+        reveal_type(x)  # revealed: Literal["test"]
+    else:
+        x = "terminal"
+        reveal_type(x)  # revealed: Literal["terminal"]
+        raise ValueError
+    reveal_type(x)  # revealed: Literal["test"]
+```
+
+In `f`, we should be able to determine that the `else` branch ends in a terminal statement, and that
+the `return` statement can only be executed when the condition is true. We should therefore consider
+the reference always bound, even though `x` is only bound in the true branch.
+
+Similarly, in `g`, we should see that the assignment of the value `"terminal"` can never be seen by
+the final `reveal_type`.
+
+## `return`
+
+A `return` statement is terminal; bindings that occur before it are not visible after it.
+
+```py
+def resolved_reference(cond: bool) -> str:
+    if cond:
+        x = "test"
+    else:
+        return "early"
+    return x  # no possibly-unresolved-reference diagnostic!
+
+def return_in_then_branch(cond: bool):
+    if cond:
+        x = "terminal"
+        reveal_type(x)  # revealed: Literal["terminal"]
+        return
+    else:
+        x = "test"
+        reveal_type(x)  # revealed: Literal["test"]
+    reveal_type(x)  # revealed: Literal["test"]
+
+def return_in_else_branch(cond: bool):
+    if cond:
+        x = "test"
+        reveal_type(x)  # revealed: Literal["test"]
+    else:
+        x = "terminal"
+        reveal_type(x)  # revealed: Literal["terminal"]
+        return
+    reveal_type(x)  # revealed: Literal["test"]
+
+def return_in_both_branches(cond: bool):
+    if cond:
+        x = "terminal1"
+        reveal_type(x)  # revealed: Literal["terminal1"]
+        return
+    else:
+        x = "terminal2"
+        reveal_type(x)  # revealed: Literal["terminal2"]
+        return
+
+def return_in_try(cond: bool):
+    x = "before"
+    try:
+        if cond:
+            x = "test"
+            return
+    except:
+        # TODO: Literal["before"]
+        reveal_type(x)  # revealed: Literal["before", "test"]
+    else:
+        reveal_type(x)  # revealed: Literal["before"]
+    finally:
+        reveal_type(x)  # revealed: Literal["before", "test"]
+    reveal_type(x)  # revealed: Literal["before", "test"]
```

Comment:
Here we're considering the `"test"` assignment visible even though we return immediately after.  Is that because we assume an exception might occur after the assignment but before the `return`?  (It looks like we take a flow snapshot after every definition in the `try` body.  Should we instead take snapshots after each statement that could in theory raise an exception?  i.e. actual `raise`s plus anything that desugars into a call?)

Reply:
In practice almost everything in Python desugars into a call, so currently our over-approximation is to take a snapshot after every definition. But yes, I think in future we should aim to refine that to better handle cases where an exception really isn't possible.

Reply:
theoretically even a statement like `x = 3` _could_ raise an exception, since there's always the possibility of a `KeyboardInterrupt` terminating the current scope. But we probably want to pretend that `KeyboardInterrupt` doesn't exist, since most code pretends that `KeyboardInterrupt` doesn't exist, and it will be annoying for users if we explain to them that we're complaining that some variable here might be undefined because they haven't accounted for the possibility of a `KeyboardInterrupt` in their code.

TL;DR: I agree with Carl. We should add a more sophisticated analysis of exactly which statements are actually capable of raising exceptions at some point. We just haven't gotten to it yet.

Reply:
:+1: Thanks for the details!  I was mostly wanting to confirm my understanding of the result and that there wasn't a bug I needed to fix.